src/math.rs

```diff
@@ -1 +1,201 @@
+use crate::internal_math;
 
+use std::mem::swap;
+
+#[allow(clippy::many_single_char_names)]
+pub fn pow_mod(x: i64, mut n: i64, m: u32) -> u32 {
+    assert!(0 <= n && 1 <= m && m <= 2u32.pow(31));
+    if m == 1 {
+        return 0;
+    }
+    let bt = internal_math::Barrett::new(m);
+    let mut r = 1;
+    let mut y = internal_math::safe_mod(x, m as i64) as u32;
+    while n != 0 {
+        if n & 1 != 0 {
+            r = bt.mul(r, y);
+        }
+        y = bt.mul(y, y);
+        n >>= 1;
+    }
+    r
+}
+
+pub fn inv_mod(x: i64, m: i64) -> i64 {
+    assert!(1 <= m);
+    let z = internal_math::inv_gcd(x, m);
+    assert!(z.0 == 1);
+    z.1
+}
+
+pub fn crt(r: &[i64], m: &[i64]) -> (i64, i64) {
+    assert!(r.len() == m.len());
```

Comment:
nit: `assert_eq!` is slightly better because if assertion fails, `assert_eq!` displays the compared values.
```suggestion
    assert_eq!(r.len(), m.len());
```